### PR TITLE
Pass through parameters on Range tasks to of tasks

### PR DESCRIPTION
**Pull request by nugend**
## Description
I've added "of_params" as an additional parameter on the RangeBase class. The approach is based on discussion with @Tarrasch 

## Motivation and Context
Please see issue from #1627 

## Have you tested this? If so, how?
I have included unit tests.

I have not yet tested this in our system, but I will do so if the maintainers think the approach used is fine and report back on the behavior.

### CODE REVIEW

**Comment by nugend** [comment]
Actually, I don't think bulk_complete can work correctly with this change. You'd have to pass in the arguments both explicitly and using the of_params field. I'm working on a fix, but I don't know if there's a way to preserve the existing interface of bulk_complete. The problem being that RangeDaily and RangeHourly were both providing a list of parameters to those functions rather than a complete argument tuple.

The way I'm approaching a change to this is to pass a list of dictionaries. If there's a way to make python use a single value as both the positional and keyword arguments for a function, that'd be nice, but I'm not aware of it.

**Comment by nugend** [comment]
Actually, I see that MixinBulkComplete is set to handle the case that the parameter tuples can be any of a single value, a positional argument tuple, or a keyword argument tuple.

However, the BigQuery code from contribs doesn't expect this behavior.

**Comment by nugend** [comment]
Found one more small issue. Seems to be working correctly in our system now though.

**Comment by Tarrasch** [comment]
Docs are missing my friend. ^^

They can either be added in the code module or in the docs which mention range (you can `git grep Range` for it).

**Comment by nugend** [comment]
Sorry. Hope that addresses the docs issue or is there a specific area that needs to be addressed?

File: doc/luigi_patterns.rst [hunk 1]
```diff
@@ -102,10 +102,17 @@ and stop (exclusive) parameters specified:
...
+The Range classes provide a parameter which accepts a :class:`~luigi.params.DictParameter`
+and passes any parameters onwards for this purpose.
+
+.. code-block:: console
+
+	luigi RangeDaily --of MyTask --start 2014-10-31 --of-param '{"my-param": 123}'
+
+Alternatively, you can specify parameters at the task family level (as described :ref:`here <Parameter-class-level-parameters>`),
+however these will not appear in the task name for the upstream Range task which
+can have implications in how the scheduler and visualizer handle task instances.
```
Comment (Tarrasch) [inline doc/luigi_patterns.rst]: Very well written. Thanks! :)

**Comment by Tarrasch** [comment]
A bit unfortunate that we have to resort to code with `__func__` and stuff. And actually `bulk_complete` won't be a `@classmethod` any more.

Other than that I'm fine with this. @ulzha, do you want to take a quick look? Other than that I'll merge day this in a day or two.

**Comment by Tarrasch** [comment]
Actually let's not merge this until we've resolved https://github.com/spotify/luigi/issues/1657#issuecomment-216918636. In particular it seems like execution summary is broken for anything using DictParameter.

**Comment by nugend** [comment]
Ah, I noticed that earlier but thought it must have been some other bug on my end. Since all the tests passed, I figured it was fine. Makes sense to wait for a fix on that before putting in DictParameters retroactively. Please let me know when a solution is available so I can perform refactoring as necessary. I'll be using this solution for the time being since I actually don't have an issue if these error out after completing. Will update if I encounter any weirdness.

I considered changing the interface of bulk_complete, but figured it'd be better to make it a little weird for now and not significantly break the expected API. Then later at a point where you are comfortable making a backwards incompatible change, move it to a staticmethod.

**Comment by Tarrasch** [comment]
There's no need for you to change your implementation once that bug-fix is done. You can just sit back and relax for now. ^^

Yea. I thing you did the right thing regarding the `bulk_complete`. But perhaps you should add an implementation note that one "must" use the `cls` variable to instantiate the class and not use `YourTaskName` as a constructor?

**Comment by nugend** [comment]
Should that be in the docs for Range or for Task?

**Comment by Tarrasch** [comment]
@nugend I guess for Task right? Because when somebody looks at how to implement `bulk_complete`, they're looking at the documentation for Tasks in general.

**Comment by nugend** [comment]
The reason I asked was because nothing actually changed in Task and any other usage of bulk_complete is likely custom created (and likely does not have any sort of similar issue).

Would it make sense to add bulk_complete to RangeBase so that it could be documented inside the Range classes themselves?

**Comment by Tarrasch** [comment]
I don't think it makes sense to add `bulk_complete` to RangeBase. Because it's supposed to be overridden by classes that aren't subclasses of RangeBase. Right?

**Comment by nugend** [comment]
Yes, sorry. I didn't yet have my coffee when I replied earlier. What I mean is just that this isn't *really* a change in how you use bulk_complete. It's a change in how bulk_complete is invoked when used with a RangeBase derived class. Strictly speaking, there could be some other user implemented WrapperTask which uses bulk_complete to calculate the required child tasks and *that* usage of bulk_complete wouldn't change and the ability to use MixinNaiveBulkComplete would support that just as happily.

To the larger point, depending on how it is decided that bulk_complete should change, maybe just more thorough documentation of what the possible types parameter_tuples can inhabit would be sufficient for ensuring correct usage. Or possibly, a helper to generate the tasks associated with each parameter_tuple

**Comment by Tarrasch** [comment]
Ok. Let's wait with the docs change. This PR is still blocked on #1657 though. :)

**Pull request merged**